packages/web-features-figuring/src/baseline/index.ts

```diff
@@ -0,0 +1,244 @@
+  let baseline_high_date = null;
+  if (isBaselineLow) {
+    const initialReleases = [...s.entries()].map(([, r]) => r);
+    const keystoneRelease = initialReleases
+      .sort((a, b) => {
+        assert(a !== undefined);
+        assert(b !== undefined);
+        return Temporal.PlainDate.compare(
+          a.date() ?? VERY_FAR_FUTURE_DATE,
+          b.date() ?? VERY_FAR_FUTURE_DATE,
+        );
+      })
+      .pop();
+
+    assert(keystoneRelease instanceof Release);
+    baseline_low_date = (keystoneRelease.date() ?? VERY_FAR_FUTURE_DATE)
+      .toString()
+      .slice(0, 10);
+
+    if (isBaselineHigh) {
+      const date = keystoneRelease.date();
+      assert(date !== null);
+      baseline_high_date = date
+        .add(BASELINE_LOW_TO_HIGH_DURATION)
+        .toString()
+        .slice(0, 10);
+    }
+  }
+
+  return {
+    compatKey,
+    baseline,
+    baseline_low_date,
+    baseline_high_date,
+    support: s,
+    toJSON: function () {
+      return jsonify(this);
+    },
+  };
+}
+
+function withAncestors(compatKey: string, compat: Compat): string[] {
+  const items = compatKey.split(".");
+  const ancestors: string[] = [];
+
+  let current = items.shift();
+  while (items.length) {
+    current = `${current}.${items.shift()}`;
+
+    const data = compat.query(current);
+    if (typeof data === "object" && data !== null && "__compat" in data) {
+      ancestors.push(current);
+    }
+  }
+  return ancestors;
+}
+
+function minStatus(statuses: SupportStatus[]): "high" | "low" | false {
+  let bestOf: "high" | "low" = "high";
+  for (const { baseline } of statuses) {
+    if (baseline === false) {
+      return false;
+    }
+    if (baseline === "low") {
+      bestOf = "low";
+    }
+    if (baseline === "high" && bestOf === "high") {
+      bestOf = "high";
+    }
+  }
+  return bestOf;
+}
+
+function minSupport(
+  supports: Map<Browser, Release | undefined>[],
+): Map<Browser, Release | undefined> {
+  const collated = new Map<Browser, (Release | undefined)[]>();
+
+  for (const support of supports) {
+    for (const [browser, release] of support) {
+      collated.set(browser, [...(collated.get(browser) ?? []), release]);
+    }
+  }
+
+  const support: Map<Browser, Release | undefined> = new Map();
+  for (const [browser, releases] of collated) {
+    if (releases.includes(undefined)) {
+      support.set(browser, undefined);
+    } else {
+      support.set(
+        browser,
+        releases
+          .sort((r1, r2) => (r1 as Release).compare(r2 as Release))
+          .at(-1),
+      );
+    }
+  }
+  return support;
+}
+
+function latestDate(dates: (string | null)[]): string | null {
+  if (dates.some((date) => date === null)) {
+    return null;
+  }
+
+  const sorted = dates
+    .map((date) => {
+      assert(date !== null);
+      return Temporal.PlainDate.from(date);
+    })
+    .sort(Temporal.PlainDate.compare);
+
+  assert(sorted.length >= 1);
+
+  return sorted.at(-1)?.toString() ?? null;
```

Comment:
This works, and I'm sure it's fast enough, but it can be both simpler and faster to just keep track of a min/max in one loop, O(n) instead of O(n log(n)). Since dates are strings of the same format, it actually works fine to compare then as strings too. What I'd suggest:

```suggestion
  let latest = null;
  for (const date of dates) {
    if (date === null) {
      return null;
    }
    assert(date.length === 10);
    if (!latest || date > latest) {
      latest = date;
    }
  }
  return latest;
```